x/claim/keeper/claim.go

```diff
@@ -0,0 +1,551 @@
+	authvestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/Stride-Labs/stride/utils"
+	"github.com/Stride-Labs/stride/x/claim/types"
+	vestingtypes "github.com/Stride-Labs/stride/x/claim/vesting/types"
+	epochstypes "github.com/Stride-Labs/stride/x/epochs/types"
+)
+
+func (k Keeper) LoadAllocationData(ctx sdk.Context, allocationData string) bool {
+	records := []types.ClaimRecord{}
+
+	lines := strings.Split(allocationData, "\n")
+	allocatedFlags := map[string]bool{}
+	for _, line := range lines {
+		data := strings.Split(line, ",")
+		if data[0] == "" || data[1] == "" || data[2] == "" {
+			continue
+		}
+
+		airdropIdentifier := data[0]
+		sourceChainAddr := data[1]
+		airdropWeight := data[2]
+		strideAddr := utils.ConvertAddressToStrideAddress(sourceChainAddr)
+		if strideAddr == "" {
+			continue
+		}
+		allocationIdentifier := airdropIdentifier + strideAddr
+
+		// Round weight value so that it always has 10 decimal places
+		weightFloat64, err := strconv.ParseFloat(airdropWeight, 64)
+		if err != nil {
+			continue
+		}
+
+		weightStr := fmt.Sprintf("%.10f", weightFloat64)
+		weight, err := sdk.NewDecFromStr(weightStr)
+		if weight.IsNegative() || weight.IsZero() {
+			continue
+		}
+
+		if err != nil || allocatedFlags[allocationIdentifier] {
+			continue
+		}
+
+		_, err = sdk.AccAddressFromBech32(strideAddr)
+		if err != nil {
+			continue
+		}
+
+		records = append(records, types.ClaimRecord{
+			AirdropIdentifier: airdropIdentifier,
+			Address:           strideAddr,
+			Weight:            weight,
+			ActionCompleted:   []bool{false, false, false},
+		})
+
+		allocatedFlags[allocationIdentifier] = true
+	}
+
+	k.SetClaimRecordsWithWeights(ctx, records)
+	return true
+}
+
+// Remove duplicated airdrops for given params
+func (k Keeper) GetUnAllocatedUsers(ctx sdk.Context, identifier string, users []string, weights []sdk.Dec) ([]string, []sdk.Dec) {
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, append([]byte(types.ClaimRecordsStorePrefix), []byte(identifier)...))
+	newUsers := []string{}
+	newWeights := []sdk.Dec{}
+	for idx, user := range users {
+		strideAddr := utils.ConvertAddressToStrideAddress(user)
+		addr, _ := sdk.AccAddressFromBech32(strideAddr)
+		// If new user, then append user and weight
+		if !prefixStore.Has(addr) {
+			newUsers = append(newUsers, user)
+			newWeights = append(newWeights, weights[idx])
+		}
+	}
+
+	return newUsers, newWeights
+}
+
+// Get airdrop duration for action
+func GetAirdropDurationForAction(action types.Action) int64 {
+	if action == types.ActionDelegateStake {
+		return int64(types.DefaultVestingDurationForDelegateStake.Seconds())
+	} else if action == types.ActionLiquidStake {
+		return int64(types.DefaultVestingDurationForLiquidStake.Seconds())
+	}
+	return int64(0)
+}
+
+// Get airdrop by distributor
+func (k Keeper) GetAirdropByDistributor(ctx sdk.Context, distributor string) *types.Airdrop {
```

Comment:
Why do we need to fetch by distributor address?

Reply:
Basically, creating multiple airdrops with same distributor is not allowed.
So when the tx msg for creating airdrop is proceeded, we check if there is already an airdrop running with given distributor address and if exist, then tx msg is supposed to be reverted.

Reply:
Got it, makes sense. Can the distributor address be refilled with more tokens?

Reply:
Yes